cpp/src/arrow/util/iterator_test.cc

```diff
@@ -214,6 +255,270 @@ TEST(TestVectorIterator, RangeForLoop) {
+    current_repeat++;
+    bool ready_for_next = false;
+    if (current_repeat == repeat_count) {
+      current_repeat = 0;
+      ready_for_next = true;
+    }
+    return TransformYield(next, ready_for_next);
+  };
+}
+
+TEST(TestIteratorTransform, Repeating) {
+  auto original = VectorIt({1, 2, 3});
+  auto repeated = MakeTransformedIterator<TestInt, TestInt>(std::move(original),
+                                                            MakeRepeatN<TestInt>(2));
+  AssertIteratorMatch({1, 1, 2, 2, 3, 3}, std::move(repeated));
+}
+
+template <typename T>
+Transformer<T, T> MakeFilter(std::function<bool(T&)> filter) {
+  return [filter](T next) -> Result<TransformFlow<T>> {
+    if (filter(next)) {
+      return TransformYield(next);
+    } else {
+      return TransformSkip();
+    }
+  };
+}
+
+template <typename T>
+Transformer<T, T> MakeAbortOnSecond() {
+  int counter = 0;
+  return [counter](T next) mutable -> Result<TransformFlow<T>> {
+    if (counter++ == 1) {
+      return Status::Invalid("X");
+    }
+    return TransformYield(next);
+  };
+}
+
+TEST(TestIteratorTransform, SkipSome) {
+  // Exercises TransformSkip
+  auto original = VectorIt({1, 2, 3});
+  auto filter = MakeFilter<TestInt>([](TestInt& t) { return t.value != 2; });
+  auto filtered = MakeTransformedIterator(std::move(original), filter);
+  AssertIteratorMatch({1, 3}, std::move(filtered));
+}
+
+TEST(TestIteratorTransform, SkipAll) {
+  // Exercises TransformSkip
+  auto original = VectorIt({1, 2, 3});
+  auto filter = MakeFilter<TestInt>([](TestInt& t) { return false; });
+  auto filtered = MakeTransformedIterator(std::move(original), filter);
+  AssertIteratorMatch({}, std::move(filtered));
+}
+
+TEST(TestIteratorTransform, Abort) {
+  auto original = VectorIt({1, 2, 3});
+  auto transformed =
+      MakeTransformedIterator(std::move(original), MakeAbortOnSecond<TestInt>());
+  ASSERT_OK(transformed.Next());
+  ASSERT_RAISES(Invalid, transformed.Next());
+}
+
+TEST(TestAsyncIteratorTransform, SkipSome) {
+  auto original = AsyncVectorIt({1, 2, 3});
+  auto filter = MakeFilter<TestInt>([](TestInt& t) { return t.value != 2; });
+  auto filtered = TransformAsyncGenerator(std::move(original), filter);
+  AssertAsyncGeneratorMatch({1, 3}, std::move(filtered));
+}
+
+TEST(TestAsyncUtil, ReadaheadFailed) {
+  auto source = []() -> Future<TestInt> {
+    return Future<TestInt>::MakeFinished(Status::Invalid("X"));
+  };
+  auto readahead = AddReadahead<TestInt>(source, 10);
+  auto next = readahead();
+  ASSERT_EQ(Status::Invalid("X"), next.status());
+}
+
+TEST(TestAsyncUtil, Readahead) {
+  int num_delivered = 0;
+  auto source = [&num_delivered]() {
+    if (num_delivered < 5) {
+      return Future<TestInt>::MakeFinished(num_delivered++);
+    } else {
+      return Future<TestInt>::MakeFinished(IterationTraits<TestInt>::End());
+    }
+  };
+  auto readahead = AddReadahead<TestInt>(source, 10);
+  // Should not pump until first item requested
```

Comment:
This is indeed different than the synchronous implementation.  I don't think the synchronous implementation is correct.  For example, if I create a `ThreadedTableReader` then it will start reading from the source file BEFORE I call `reader->Read()`.  I find this odd and not the desired behavior.

The AsyncThreadedTableReader will not start reading from the file until the caller calls `reader->Read()` but it will create the background thread.  Once we make the filesystem interfaces async then ideally it won't even create a background thread or do any work other than instantiation.

Reply:
Reading from the source file before `Read()` is called is by design, so that the reading can start in the background without the caller having to wait.

You're right that for an async reader it's much less useful, though.

Reply:
I'm not sure that it is less useful for an async reader.  It isn't a matter of blocking either.  The threaded table reader could have the same behavior.  Also, the serial table reader was already different from the threaded table reader in this regard (the async reader mimics the serial reader)...

```
  ARROW_ASSIGN_OR_RAISE(auto csv_reader, arrow::csv::TableReader::Make(...);  // Should the read start here?
  ARROW_ASSIGN_OR_RAISE(auto table, csv_reader->Read());                                 // Or here?
```
As a user I really wouldn't expect the file to start getting read until I called `Read()`.  I agree that it shouldn't have much impact though and I'd be content to change it.